### PR TITLE
Abstract the Postgres management client to allow both single and flexible server

### PR DESCRIPTION
**Pull request by tonybaloney**
- [x] Implement server listing
- [x] Implement connection string
- [x] Implement database list
- [x] Implement provisioning

### CODE REVIEW

**Comment by tonybaloney** [comment]
screenshot 2021-06-18 at 11 05 04
server listing (both single and flexible)

**Comment by tonybaloney** [comment]
The provisioning for flexible creates the server but then hits this issue-- 
```
3:34:09 PM: Creating PostgreSQL Server "ant-flexi-2"... It should be ready in several minutes.
3:34:44 PM: Error: The resource type 'locations/azureAsyncOperation' could not be found in the namespace 'Microsoft.DBforPostgreSQL' for api version '2021-04-10-privatepreview'. The supported api-versions are '2017-12-01-preview,2017-12-01,2018-03-29-privatepreview,2020-10-05-privatepreview,2020-02-14-privatepreview,2020-02-14-preview,2020-11-05-preview'.
```

**Comment by tonybaloney** [comment]
Provisioning for single server always fails saying the password is too weak so I think the call mask is causing it to be replaced with *****

**Comment by tonybaloney** [comment]
Again for deleting single servers

```
The resource type 'locations/azureAsyncOperation' could not be found in the namespace 'Microsoft.DBforPostgreSQL' for api version '2021-04-10-privatepreview'. The supported api-versions are '2017-12-01-preview,2017-12-01,2018-03-29-privatepreview,2020-10-05-privatepreview,2020-02-14-privatepreview,2020-02-14-preview,2020-11-05-preview'.
```

File: src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts
Comment (wwlorey) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts:159]: Is there any info you could add to the description that would account for why "ms" and "s" are included in the labels? Or is this common SKU naming that people will be familiar with? It just looks a little strange to me
Screen Shot 2021-06-22 at 12 06 46 PM

File: src/postgres/abstract/AbstractPostgresClient.ts
Comment (wwlorey) [inline src/postgres/abstract/AbstractPostgresClient.ts:14]: I'd be interested to hear what @microsoft/vscodeazuretoolsdev thinks about this but I think we can get rid of `IAbstractPostgresClient` and let this class stand on its own. The interface is useful if we're creating multiple classes that implement it but here it's only used with a single class. And the only other places where it's used are typings for `createAzureClient` calls which can all be replaced with type `AbstractPostgresClient`

File: src/tree/SubscriptionTreeItem.ts
Comment (wwlorey) [inline src/tree/SubscriptionTreeItem.ts:40]:
```suggestion
        const postgresServers: PostgresAbstractServerList = await postgresClient.listServers();
```

File: src/postgres/abstract/AbstractPostgresClient.ts
Comment (wwlorey) [inline src/postgres/abstract/AbstractPostgresClient.ts:67]:
```suggestion
    async listAllServers(): Promise<Models.PostgresAbstractServerList> {
```

It doesn't have to be this exact name but I want to convey that it's listing flexible and single servers. And it's the only function in this class that doesn't just do one or the other

File: src/postgres/abstract/AbstractPostgresClient.ts
Comment (wwlorey) [inline src/postgres/abstract/AbstractPostgresClient.ts:34]:
```suggestion
    async createOrUpdateFirewallRule(serverType: Models.PostgresServerType, resourceGroup: string, name: string, ruleName: string, rule: Models.AbstractFirewallRule): Promise<Models.AbstractFirewallRule> {
```

File: src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts
Comment (wwlorey) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts:50]: nit: not sure why those parentheses were there before
```suggestion
    const availability: AbstractNameAvailability = await client.checkNameAvailability(serverType, name);
```

File: src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts
Comment (wwlorey) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts:156]: Could you include a commented link to the source for this info?

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCreateStep.ts
Comment (wwlorey) [inline src/tree/CosmosDBAccountWizard/CosmosDBAccountCreateStep.ts:33]:
```suggestion
            kind: defaultExperience.kind,
```

I don't think this is needed either. Was typescript complaining about this?

**Review by wwlorey — COMMENTED** [review]
> The provisioning for flexible creates the server but then hits this issue--
> 
> ```
> 3:34:09 PM: Creating PostgreSQL Server "ant-flexi-2"... It should be ready in several minutes.
> 3:34:44 PM: Error: The resource type 'locations/azureAsyncOperation' could not be found in the namespace 'Microsoft.DBforPostgreSQL' for api version '2021-04-10-privatepreview'. The supported api-versions are '2017-12-01-preview,2017-12-01,2018-03-29-privatepreview,2020-10-05-privatepreview,2020-02-14-privatepreview,2020-02-14-preview,2020-11-05-preview'.
> ```

I'm running into this as well. Do you know if/when that api version will be supported?

File: src/postgres/abstract/AbstractPostgresClient.ts
Comment (neelip) [reply src/postgres/abstract/AbstractPostgresClient.ts:14]: I agree with Will on this. Right now, IAbstractPostgresClient is only used for declaring type. I think it is a good idea to remove it and just use AbstractPostgresClient, I don't see any harm in it.

File: src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts
Comment (tonybaloney) [reply src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts:160]: This is how they're documented in the official docs https://docs.microsoft.com/en-au/azure/postgresql/flexible-server/concepts-compute-storage#compute-tiers-vcores-and-server-types

I don't know what m or s means.

**Comment by tonybaloney** [comment]
@wwlorey updates completed from the review

**Comment by wwlorey** [comment]
This looks good 👍👍  I'll wait until we hear back about the API problem before testing again & getting this merged

**Comment by tonybaloney** [comment]
> This looks good 👍👍  I'll wait until we hear back about the API problem before testing again & getting this merged

There is an issue with the SDK release. I'm waiting for a patch and will update packages.json once it's done

File: src/postgres/abstract/AbstractPostgresClient.ts
Comment (ejizba) [inline src/postgres/abstract/AbstractPostgresClient.ts:33]: The vast majority of these methods are just copy/pasted for both clients. Rather than duplicating the code, how about we do something like this:

```typescript
export type AbstractPostgresClient = PostgreSQLFlexibleManagementClient | PostgreSQLManagementClient;

export function createAbstractPostgresClient(serverType: PostgresServerType, clientInfo: ISubscriptionContext): AbstractPostgresClient {
    switch (serverType) {
        case PostgresServerType.Flexible:
            return createAzureClient(clientInfo, PostgreSQLFlexibleManagementClient);
        case PostgresServerType.Single:
            return createAzureClient(clientInfo, PostgreSQLManagementClient);
        default:
            throw new Error("Service not implemented.");
    }
}
```

As an example, `configurePostgresFirewall.ts` would then look like this:
```typescript
const client = createAbstractPostgresClient(treeItem.serverType, treeItem.root);
await client.firewallRules.createOrUpdate(resourceGroup, serverName, firewallRuleName, {
    startIpAddress: ip,
    endIpAddress: ip
});
```

File: src/postgres/abstract/AbstractPostgresClient.ts
Comment (ejizba) [inline src/postgres/abstract/AbstractPostgresClient.ts:44]: `createServer` and `listAllServers` are the only two methods on this class that aren't simple copy/paste and thus the only reason we _might_ need this class still. However, since each method is only used once, I don't think there's much value to putting them in a separate class like this. The true value of abstracting it away would be if we were calling the methods multiple times, which I don't see happening. I say we go with my `createAbstractPostgresClient` function above and move this code directly into `PostgresServerCreateStep`

File: src/postgres/abstract/maps.ts
Comment (ejizba) [inline src/postgres/abstract/maps.ts:12]: Rather than copy/pasting the properties like this, we can directly use the original object and a `type` instead of an `interface`.

The only complication with `PostgresAbstractServer` is `serverType`. Do we actually need that property or is there already some property that identifies if it's flexible/single? If we _do_ need the property, you can still use the original object, but just add `serverType` on top like this:
```typescript
export type PostgresAbstractServer = (SingleModels.Server | FlexibleModels.Server) & { serverType: PostgresServerType; }
```

```typescript
export function singleAsAbstractServer(server: SingleModels.Server): PostgresAbstractServer {
    return Object.assign(server, { serverType: PostgresServerType.Single });
}
```

File: src/postgres/abstract/models.ts
Comment (ejizba) [inline src/postgres/abstract/models.ts:59]:
```
export type PostgresAbstractDatabase = FlexibleModels.Database | SingleModels.Database;
```

File: src/postgres/abstract/models.ts
Comment (ejizba) [inline src/postgres/abstract/models.ts:87]:
```
tier?: FlexibleModels.SkuTier | SingleModels.SkuTier;
```

File: src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts
Comment (ejizba) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts:53]: If I follow my suggested pattern for `type` vs. `interface`, it'll look something like this:
```typescript
export type AbstractNameAvailability = SingleModels.NameAvailability | FlexibleModels.NameAvailability;
```

However, all of the sudden this line starts complaining because `reason` doesn't exist on `FlexibleModels.NameAvailability`. That implies to me that either:
1. The types on `FlexibleModels.NameAvailability` are wrong
2. The types are correct and this logic doesn't actually work for flexible servers

If 2 is true, we need to update this logic. If 1 is true, we can update the type like so:
```typescript
export type AbstractNameAvailability = SingleModels.NameAvailability | (FlexibleModels.NameAvailability & {
    /**
    * NOTE: The types on `FlexibleModels.NameAvailability` are incorrect, this property does exist
    */
    reason?: string;
})
```

File: src/postgres/abstract/maps.ts
Comment (ejizba) [inline src/postgres/abstract/maps.ts:64]: If a function is only used once, we normally put it in the file where it's used, rather than in a generic file like "maps.ts" (unless we predict it being used multiple times in the future, but I don't see that happening here). I think `asFlexibleParameters` and `asSingleParameters` would fit better in `PostgresServerCreateStep.ts`

**Review by ejizba — COMMENTED** [review]
Sorry for jumping in here a little late, but I think there's too much duplicated code inside `AbstractPostgresClient` and `models.ts`. We lose a bit of type safety and risk some silly copy/paste bugs. We should be able to leverage TypeScript features like union types more heavily to clean it up

**Comment by tonybaloney** [comment]
> Sorry for jumping in here a little late, but I think there's too much duplicated code inside `AbstractPostgresClient` and `models.ts`. We lose a bit of type safety and risk some silly copy/paste bugs. We should be able to leverage TypeScript features like union types more heavily to clean it up

@ejizba thanks for the feedback. My TS knowledge is still pretty basic. This PR is still blocked on an SDK patch so I'll work through your feedback and update this PR.

File: src/postgres/abstract/maps.ts
Comment (tonybaloney) [reply src/postgres/abstract/maps.ts:12]: The models are very different. Doing this is throwing a range of compiler errors

```
Argument of type '(server: Server) => PostgresAbstractServer' is not assignable to parameter of type '((value: Server, index: number, array: Server[]) => PostgresAbstractServer) & ((value: Server, index: number, array: Server[]) => PostgresAbstractServer)'.
  Type '(server: Server) => PostgresAbstractServer' is not assignable to type '(value: Server, index: number, array: Server[]) => PostgresAbstractServer'.
    Types of parameters 'server' and 'value' are incompatible.
      Type 'import("/Users/anthonyshaw/projects/vscode-cosmosdb/node_modules/@azure/arm-postgresql/esm/models/index").Server' is not assignable to type 'import("/Users/anthonyshaw/projects/vscode-cosmosdb/node_modules/@azure/arm-postgresql-flexible/esm/models/index").Server'.
        Types of property 'sku' are incompatible.
          Type 'import("/Users/anthonyshaw/projects/vscode-cosmosdb/node_modules/@azure/arm-postgresql/esm/models/index").Sku | undefined' is not assignable to type 'import("/Users/anthonyshaw/projects/vscode-cosmosdb/node_modules/@azure/arm-postgresql-flexible/esm/models/index").Sku | undefined'.
            Type 'import("/Users/anthonyshaw/projects/vscode-cosmosdb/node_modules/@azure/arm-postgresql/esm/models/index").Sku' is not assignable to type 'import("/Users/anthonyshaw/projects/vscode-cosmosdb/node_modules/@azure/arm-postgresql-flexible/esm/models/index").Sku'.
              Types of property 'tier' are incompatible.
                Type 'SkuTier | undefined' is not assignable to type 'SkuTier'.
                  Type 'undefined' is not assignable to type 'SkuTier'.ts(2345)
```

File: src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts
Comment (tonybaloney) [reply src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts:53]: thats just because I referred to the reason instead of the message, this can be unionised.

**Comment by tonybaloney** [comment]
@ejizba refactored as per your feedback in https://github.com/microsoft/vscode-cosmosdb/pull/1872/commits/bb01e7a8c287c2bc02309a8a423db06cb83572f9

File: src/postgres/abstract/models.ts
Comment (ejizba) [inline src/postgres/abstract/models.ts:16]: Remove these "List" types, they're very simple and I don't think they're used

File: src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts
Comment (ejizba) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts:53]: I don't think `message` is right - this is what I'm seeing
Screen Shot 2021-06-30 at 1 12 03 PM

That's for single server. Not sure what flexible looks like because of the sdk issue.

File: src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts
Comment (ejizba) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerSkuStep.ts:45]: Always use `===` instead of `==`

I'm kind of surprised our linting doesn't catch this - thoughts @wwlorey
https://eslint.org/docs/rules/eqeqeq

File: src/postgres/abstract/models.ts
Comment (tonybaloney) [reply src/postgres/abstract/models.ts]: I've seen this pattern a few times in this module, I'll raise another PR with other occurrences.

**Comment by tonybaloney** [comment]
Blocked by pending update to https://github.com/Azure/azure-sdk-for-js/issues/15979

**Comment by tonybaloney** [comment]
Related https://github.com/Azure/azure-sdk-for-js/pull/16136

**Comment by tonybaloney** [comment]
@wwlorey @ejizba this PR is now finished/ready for review. the backend dependency on the JS SDK has been fixed and I've updated package.json with the new version

File: src/postgres/commands/createPostgresServer/steps/PostgresServerCreateStep.ts
Comment (ejizba) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerCreateStep.ts:17]: Any specific reason this default changed from 10 to 11? If we're going to update the default, I'd prefer that as a separate PR instead of hidden in this larger refactoring. I'm not super versed on postgres versions, so I'd wanna know what the portal uses as a default, what the support level is for 10 vs 11, etc.

File: src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts
Comment (ejizba) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts:50]: This isn't workin for me for flexible servers. It's silently failing under the covers because of this error:

> Unknown resource type specified for a subscription level name availability check. Supported resource types are: Microsoft.DBforPostgreSQL/flexibleServers.

File: src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts
Comment (ejizba) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerNameStep.ts:53]: Let's just return `availability.message` - not sure why we weren't using that already

File: src/postgres/commands/createPostgresServer/steps/PostgresServerCreateStep.ts
Comment (ejizba) [inline src/postgres/commands/createPostgresServer/steps/PostgresServerCreateStep.ts:19]: I did a bit of a refactor and caused some merge conflicts for ya - sorry bout that, but please fix them :)

File: src/postgres/commands/createPostgresServer/steps/PostgresServerCreateStep.ts
Comment (tonybaloney) [reply src/postgres/commands/createPostgresServer/steps/PostgresServerCreateStep.ts:17]: The portal default is 11 for single server and 12 for flexible server. 

Single server supports 9.6-11
Flexible server supports 11-13

10 is no longer supported in flexible server, but still in single server (but no longer the default as it was when this was originally written).

I can make it a separate PR but would have to replace this with a switch to default to 10 for single and 11 for flexible

Comment (ejizba) [reply src/postgres/commands/createPostgresServer/steps/PostgresServerCreateStep.ts:17]: Matching the portal would be best, aka 11 for single and 12 for flexible. I think I'm okay including this in the current PR.

File: src/AzureDBExperiences.ts
Comment (ejizba) [inline src/AzureDBExperiences.ts:114]: When I was trying out the portal experience to check the default versions I noticed they label Flexible as "Preview". We should do the same

**Comment by tonybaloney** [comment]
@ejizba amendments made in https://github.com/microsoft/vscode-cosmosdb/pull/1872/commits/cb353d473788089b30b04e7c23510df865c8eca8

**Review by ejizba — APPROVED** [review]
Looks good to me! @wwlorey were you still going to double check?

**Review by wwlorey — APPROVED** [review]
Looks good 👍

**Pull request merged**
